osbuild/sources.py

```diff
@@ -51,17 +54,66 @@ class SourceService(host.Service):
     """Source host service"""
 
     @abc.abstractmethod
-    def download(self, items, cache, options):
-        pass
+    def fetch_one(self, checksum, desc) -> None:
+        """Performs the actual fetch of an element described by its checksum and its descriptor"""
+
+    def config_download(self) -> None:
+        """Hook to set config for the download affecting all the elements, such as parallelism or special caches."""
+
+    def exists(self, checksum, _desc) -> bool:
+        """Returns True if the item to download is in cache. """
+        return os.path.isfile(f"{self.cache}/{checksum}")
+
+    # pylint: disable=[no-self-use]
+    def transform(self, checksum, desc) -> Tuple:
+        """Modify the input data before downloading. By default only transforms an item object to a Tupple."""
+        return checksum, desc
+
+    def _unpack(self, items: List[Tuple]) -> (List, List):
+        """Change a list of tuples [(hash:desc), ...] to two lists of the same sizes, [hash, ...], [desc, ...]."""
+        return map(list, zip(*items))
+
+    def _prepare(self, items) -> List[Tuple]:
+        """Prepare every item transforming them before downloading."""
+        return [self.transform(checksum, desc) for checksum, desc in items.items()]
+
+    def _filter(self, items: Dict) -> Dict:
+        """Only keep the items that aren't alredy in the cache."""
+        return {checksum: desc for checksum, desc in items.items() if not self.exists(checksum, desc)}
+
+    def download(self, items: Dict) -> None:
+        with concurrent.futures.ThreadPoolExecutor(max_workers=self.num_parallel_fetch) as executor:
+            for _ in executor.map(self.fetch_one, *self._unpack(self._prepare(self._filter(items)))):
+                pass
+
+    @property
+    @classmethod
+    @abstractmethod
+    def content_type(cls):
+        """The content type of the source."""
+
+    @property
+    def cache(self):
+        return self._cache
+
+    @cache.setter
+    def cache(self, val: str):
+        self._cache = os.path.join(val, self.content_type)
+        os.makedirs(self._cache, exist_ok=True)
+
+    @staticmethod
+    def load_items(fds):
+        with os.fdopen(fds.steal(0)) as f:
+            items = json.load(f)
+        return items
 
     def dispatch(self, method: str, args, fds):
         if method == "download":
-            with os.fdopen(fds.steal(0)) as f:
-                items = json.load(f)
-
-            r = self.download(items,
-                              args["cache"],
-                              args["options"])
-            return r, None
+            self.num_parallel_fetch = 1
```

Comment:
pylint [W0201](https://pylint.pycqa.org/en/latest/messages/warning/attribute-defined-outside-init.html?highlight=attribute-defined-outside-init))pylint [W0201](https://pylint.pycqa.org/en/latest/messages/warning/attribute-defined-outside-init.html?highlight=attribute-defined-outside-init), but this should be a class attribute. And maybe we just call it `max_workers` as is the argument for the thread pool, it is also technically more correct since ostree internally also has parallel fetches.